src/net/sourceforge/kolmafia/persistence/ConcoctionDatabase.java

```diff
@@ -3332,4 +3266,114 @@ public int hashCode() {
       return hash;
     }
   }
+
+  public static class UsableConcoctions {
+    private Map<ConcoctionType, LockableListModel<Concoction>> usableMap = new TreeMap<>();
+
+    public UsableConcoctions() {
+      for (ConcoctionType type : ConcoctionType.values()) {
+        usableMap.put(type, new LockableListModel<>());
+      }
+    }
+
+    public void fill() {
+      this.clear();
+      Collection<Concoction> allConcoctions = ConcoctionPool.concoctions();
+      for (Map.Entry<ConcoctionType, LockableListModel<Concoction>> entry :
+          this.usableMap.entrySet()) {
+        entry
+            .getValue()
+            .addAll(
+                allConcoctions.stream()
+                    .filter(c -> c.type == entry.getKey())
+                    .collect(Collectors.toList()));
+      }
+    }
+
+    private static <T extends Comparable<T>> boolean isSorted(Iterable<T> iterable) {
+      T last = null;
+      boolean sorted = true;
+      for (T current : iterable) {
+        if (last != null && last.compareTo(current) > 0) {
+          sorted = false;
+        }
+        last = current;
+      }
+      return sorted;
+    }
+
+    public int size() {
+      return this.usableMap.values().stream().mapToInt(List::size).sum();
+    }
+
+    public boolean contains(Concoction c) {
+      return this.usableMap.get(c.type).contains(c);
+    }
+
+    public LockableListModel<Concoction> get(ConcoctionType type) {
+      return this.usableMap.get(type);
+    }
+
+    public Collection<LockableListModel<Concoction>> values() {
+      return this.usableMap.values();
+    }
+
+    public void clear() {
+      this.usableMap.values().stream().forEach(List::clear);
+    }
+
+    public void add(Concoction c) {
+      LockableListModel<Concoction> list = this.usableMap.get(c.type);
+      int index = Collections.binarySearch(list, c);
+      if (index < 0) {
+        index = -index - 1;
+      }
+      list.add(index, c);
```

Comment:
if the index was positive, the item was in the list, so we add it again? Is that right?

Reply:
Fixed to update instead of add if found.